File: tests/test_endpoints.js
#!/usr/bin/env node
/**
 * CincoDeBio Endpoint Tests (JavaScript / Node.js)
 *
 * Usage:
 *   node test_endpoints.js
 *
 * Requires port-forwards:
 *   kubectl port-forward svc/sib-manager 8081:80 &
 *   kubectl port-forward svc/execution-api 8082:80 &
 *   kubectl port-forward svc/service-api 8084:80 &
 */

const SIB_MANAGER = process.env.SIB_MANAGER_URL || "http://localhost:8081";
const EXECUTION_API = process.env.EXECUTION_API_URL || "http://localhost:8082";
const SERVICE_API = process.env.SERVICE_API_URL || "http://localhost:8084";

let passed = 0;
let failed = 0;

async function test(name, fn) {
  try {
    await fn();
    passed++;
    console.log(`  PASS: ${name}`);
  } catch (e) {
    failed++;
    console.log(`  FAIL: ${name} — ${e.message}`);
  }
}

function assert(cond, msg) {
  if (!cond) throw new Error(msg);
}

async function fetchJSON(url, opts) {
  const res = await fetch(url, { signal: AbortSignal.timeout(10000), ...opts });
  return { res, data: await res.json() };
}

// ---- sib-manager ----

async function testSibManagerHealth() {
  const { res, data } = await fetchJSON(`${SIB_MANAGER}/health`);
  assert(res.ok, `status ${res.status}`);
  assert(data.status === "healthy", `unexpected: ${JSON.stringify(data)}`);
}

async function testGetSibMap() {
  const { res, data } = await fetchJSON(`${SIB_MANAGER}/get-sib-map`);
  assert(res.ok, `status ${res.status}`);
  assert(data.concepts, "missing concepts");
  assert(data.inputs, "missing inputs");
  assert(data.outputs, "missing outputs");
  const concepts = Object.keys(data.concepts);
  assert(concepts.length > 0, "no SIBs");
  console.log(`         SIB map: ${concepts.length} concepts, ${Object.keys(data.inputs).length} input defs, ${Object.keys(data.outputs).length} output defs`);
  for (const name of concepts.sort()) {
    const ins = data.inputs[name] || {};
    const outs = data.outputs[name] || {};
    const inTypes = (Array.isArray(ins) ? ins : Object.keys(ins)).map(t => t.includes(":") ? t.split(":")[1] : t);
    const outTypes = (Array.isArray(outs) ? outs : Object.keys(outs)).map(t => t.includes(":") ? t.split(":")[1] : t);
    console.log(`           ${name.padEnd(40)} in=[${inTypes.join(", ")}]  out=[${outTypes.join(", ")}]`);
  }
}

async function testGetInstalledSibs() {
  const { res, data } = await fetchJSON(`${SIB_MANAGER}/ext/get-installed-sibs`);
  assert(res.ok, `status ${res.status}`);
  assert(Array.isArray(data), "expected array");
  assert(data.length > 0, "no installed SIBs");
  assert(data[0].image, "missing image field");
  console.log(`         ${data.length} installed SIBs:`);
  for (const s of data) {
    const schema = s["cincodebio.schema"] || {};
    const name = schema.service_name || "?";
    const stype = (schema.service_type || "?").padEnd(12);
    const concept = schema.abstract_concept || "?";
    console.log(`           ${name.padEnd(40)} type=${stype} concept=${concept}`);
  }
}

async function testGetUninstalledSibs() {
  const { res, data } = await fetchJSON(`${SIB_MANAGER}/ext/get-uninstalled-sibs`);
  assert(res.ok, `status ${res.status}`);
  assert(Array.isArray(data), "expected array");
  console.log(`         ${data.length} uninstalled SIBs`);
}

async function testGetUtdSibFile() {
  const { res, data } = await fetchJSON(`${SIB_MANAGER}/ext/get-utd-sib-file`);
  assert(res.ok, `status ${res.status}`);
  assert(data.file, "missing file key");
  assert(data.file.length > 0, "file is empty");
  console.log(`         UTD sib file: ${data.file.length} characters`);
}

async function testCheckSibFileHashInvalid() {
  const { res, data } = await fetchJSON(`${SIB_MANAGER}/ext/check-sib-file-hash`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ fileHash: "fakehash123" }),
  });
  assert(res.ok, `status ${res.status}`);
  assert(data === "INVALID", `expected INVALID, got ${JSON.stringify(data)}`);
  console.log("         Sent fake hash → correctly returned INVALID");
}

async function testCheckSibFilesHashesEmpty() {
  // Retry once — port-forward can drop after the preceding 500 error
  for (let attempt = 0; attempt < 2; attempt++) {
    try {
      const { res, data } = await fetchJSON(`${SIB_MANAGER}/ext/check-sib-files-hashes`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ fileHashes: {} }),
      });
      assert(res.ok, `status ${res.status}`);
      assert(data.hashesValid !== undefined, `missing hashesValid: ${JSON.stringify(data)}`);
      console.log("         Sent empty hashes → response contains hashesValid");
      return;
    } catch (e) {
      if (attempt === 0) {
        await new Promise((r) => setTimeout(r, 1000));
        continue;
      }
      throw e;
    }
  }
}

async function testSibManagerState() {
  const res = await fetch(`${SIB_MANAGER}/sib-manager-state`, {
    signal: AbortSignal.timeout(10000),
  });
  assert(res.ok, `status ${res.status}`);
  const data = await res.json();
  const latest = data.latest || [];
  const installed = data.installed || [];
  const rest = data.rest || [];
  console.log(`         State: ${latest.length} latest, ${installed.length} installed, ${rest.length} pending`);
  if (rest.length > 0) {
    console.log(`         Pending: ${rest.join(", ")}`);
  }
}

// ---- execution-api ----

async function testExecutionApiHealth() {
  const { res, data } = await fetchJSON(`${EXECUTION_API}/health`);
  assert(res.ok, `status ${res.status}`);
  assert(data.status === "healthy", `unexpected: ${JSON.stringify(data)}`);
}

async function testGetWorkflows() {
  const { res, data } = await fetchJSON(`${EXECUTION_API}/ext/get-workflows`);
  assert(res.ok, `status ${res.status}`);
  assert(Array.isArray(data), "expected array");
  console.log(`         ${data.length} workflows found`);
  if (data.length > 0) {
    const statuses = {};
    data.forEach(w => { const s = w.status || "unknown"; statuses[s] = (statuses[s] || 0) + 1; });
    const statusStr = Object.entries(statuses).map(([k, v]) => `${v} ${k}`).join(", ");
    console.log(`         Statuses: ${statusStr}`);
    for (const w of data.slice(0, 5)) {
      console.log(`           id=${w.id}  owner=${w.owner || "?"}  status=${w.status || "?"}`);
    }
    if (data.length > 5) {
      console.log(`           ... and ${data.length - 5} more`);
    }
  }
}

async function testSubmitWorkflow() {
  const form = new FormData();
  const blob = new Blob(["InitTMA -> SegArrayTMA"], { type: "text/plain" });
  form.append("model", blob, "test.sibs");

  const res = await fetch(`${EXECUTION_API}/ext/model/submit?v2=false`, {
    method: "POST",
    body: form,
    signal: AbortSignal.timeout(30000),
  });
  assert(res.status === 202, `expected 202, got ${res.status}`);
  const data = await res.json().catch(() => ({}));
  console.log("         Submitted: InitTMA -> SegArrayTMA");
  if (data.url) console.log(`         Workflow URL: ${data.url}`);
}

// ---- service-api ----

async function testServiceApiHealth() {
  const { res, data } = await fetchJSON(`${SERVICE_API}/health`);
  assert(res.ok, `status ${res.status}`);
  assert(data.status === "ok" || data.status === "healthy", `unexpected: ${JSON.stringify(data)}`);
}

async function testServiceApiOpenapi() {
  const { res, data } = await fetchJSON(`${SERVICE_API}/openapi.json`);
  assert(res.ok, `status ${res.status}`);
  assert(data.paths, "missing paths");
  const paths = Object.keys(data.paths);
  assert(paths.length > 0, "no paths");
  console.log(`         ${paths.length} endpoints registered:`);
  for (const path of paths.sort()) {
    const methods = Object.keys(data.paths[path]).map(m => m.toUpperCase()).join(", ");
    console.log(`           ${methods.padEnd(6)} ${path}`);
  }
}

// ---- main ----

async function main() {
  console.log("=== sib-manager tests ===");
  await test("health", testSibManagerHealth);
  await test("get-sib-map", testGetSibMap);
  await test("get-installed-sibs", testGetInstalledSibs);
  await test("get-uninstalled-sibs", testGetUninstalledSibs);
  await test("get-utd-sib-file", testGetUtdSibFile);
  await test("check-sib-file-hash (invalid)", testCheckSibFileHashInvalid);
  await test("check-sib-files-hashes (empty)", testCheckSibFilesHashesEmpty);
  await test("sib-manager-state", testSibManagerState);

  console.log("\n=== execution-api tests ===");
  await test("health", testExecutionApiHealth);
  await test("get-workflows", testGetWorkflows);
  await test("submit-workflow", testSubmitWorkflow);

  console.log("\n=== service-api tests ===");
  await test("health", testServiceApiHealth);
  await test("openapi spec", testServiceApiOpenapi);

  console.log(`\n${"=".repeat(40)}`);
  console.log(`Results: ${passed} passed, ${failed} failed, ${passed + failed} total`);
  process.exit(failed > 0 ? 1 : 0);
}

main();

File: tests/test_endpoints.py
#!/usr/bin/env python3
"""
CincoDeBio Endpoint Tests
Tests sib-manager, execution-api, and service-api endpoints.

Usage:
    python3 test_endpoints.py [--base-url http://localhost]

Requires port-forwards:
    kubectl port-forward svc/sib-manager 8081:80 &
    kubectl port-forward svc/execution-api 8082:80 &
    kubectl port-forward svc/service-api 8084:80 &
"""

import argparse
import json
import sys
from collections import Counter
import requests

SIB_MANAGER = None
EXECUTION_API = None
SERVICE_API = None

PASSED = 0
FAILED = 0


def test(name, fn):
    global PASSED, FAILED
    try:
        fn()
        PASSED += 1
        print(f"  PASS: {name}")
    except Exception as e:
        FAILED += 1
        print(f"  FAIL: {name} — {e}")


# ---------- sib-manager ----------

def test_sib_manager_health():
    r = requests.get(f"{SIB_MANAGER}/health", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert data.get("status") == "healthy", f"unexpected: {data}"


def test_get_sib_map():
    r = requests.get(f"{SIB_MANAGER}/get-sib-map", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert "concepts" in data, "missing 'concepts' key"
    assert "inputs" in data, "missing 'inputs' key"
    assert "outputs" in data, "missing 'outputs' key"
    concepts = data["concepts"]
    assert len(concepts) > 0, "no SIBs in concepts map"
    print(f"         SIB map: {len(concepts)} concepts, {len(data['inputs'])} input defs, {len(data['outputs'])} output defs")
    for name in sorted(concepts.keys()):
        ins = data["inputs"].get(name, {})
        outs = data["outputs"].get(name, {})
        in_types = [t.split(":")[1] if ":" in t else t for t in ins] if isinstance(ins, list) else list(ins.keys())
        out_types = [t.split(":")[1] if ":" in t else t for t in outs] if isinstance(outs, list) else list(outs.keys())
        print(f"           {name:40s} in=[{', '.join(in_types)}]  out=[{', '.join(out_types)}]")


def test_get_installed_sibs():
    r = requests.get(f"{SIB_MANAGER}/ext/get-installed-sibs", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert isinstance(data, list), "expected list"
    assert len(data) > 0, "no installed SIBs"
    first = data[0]
    assert "image" in first, "SIB missing 'image' field"
    assert "cincodebio.schema" in first, "SIB missing schema"
    print(f"         {len(data)} installed SIBs:")
    for s in data:
        schema = s.get("cincodebio.schema", {})
        name = schema.get("service_name", "?")
        stype = schema.get("service_type", "?")
        concept = schema.get("abstract_concept", "?")
        print(f"           {name:40s} type={stype:12s} concept={concept}")


def test_get_uninstalled_sibs():
    r = requests.get(f"{SIB_MANAGER}/ext/get-uninstalled-sibs", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert isinstance(data, list), "expected list"
    print(f"         {len(data)} uninstalled SIBs")


def test_get_utd_sib_file():
    r = requests.get(f"{SIB_MANAGER}/ext/get-utd-sib-file", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert "file" in data, "missing 'file' key"
    assert len(data["file"]) > 0, "sib file content is empty"
    print(f"         UTD sib file: {len(data['file'])} characters")


def test_check_sib_file_hash_invalid():
    r = requests.post(
        f"{SIB_MANAGER}/ext/check-sib-file-hash",
        json={"fileHash": "fakehash123"},
        timeout=10,
    )
    assert r.status_code == 200, f"status {r.status_code}"
    assert r.json() == "INVALID", f"unexpected: {r.json()}"
    print("         Sent fake hash → correctly returned INVALID")


def test_check_sib_files_hashes_empty():
    r = requests.post(
        f"{SIB_MANAGER}/ext/check-sib-files-hashes",
        json={"fileHashes": {}},
        timeout=10,
    )
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert "hashesValid" in data, f"missing key: {data}"
    print("         Sent empty hashes → response contains hashesValid")


def test_sib_manager_state():
    r = requests.get(f"{SIB_MANAGER}/sib-manager-state", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    latest = data.get("latest", [])
    installed = data.get("installed", [])
    rest = data.get("rest", [])
    print(f"         State: {len(latest)} latest, {len(installed)} installed, {len(rest)} pending")
    if rest:
        print(f"         Pending: {', '.join(rest)}")


# ---------- execution-api ----------

def test_execution_api_health():
    r = requests.get(f"{EXECUTION_API}/health", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert data.get("status") == "healthy", f"unexpected: {data}"


def test_get_workflows():
    r = requests.get(f"{EXECUTION_API}/ext/get-workflows", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert isinstance(data, list), "expected list"
    print(f"         {len(data)} workflows found")
    if data:
        statuses = Counter(w.get("status", "unknown") for w in data)
        status_str = ", ".join(f"{v} {k}" for k, v in statuses.items())
        print(f"         Statuses: {status_str}")
        for w in data[:5]:
            print(f"           id={w['id']}  owner={w.get('owner','?')}  status={w.get('status','?')}")
        if len(data) > 5:
            print(f"           ... and {len(data)-5} more")


def test_submit_workflow():
    """Submit a minimal composed workflow model to execution-api."""
    model_content = "InitTMA -> SegArrayTMA"

    r = requests.post(
        f"{EXECUTION_API}/ext/model/submit",
        files={"model": ("test.sibs", model_content, "text/plain")},
        params={"v2": False},
        timeout=30,
    )
    assert r.status_code == 202, f"status {r.status_code}, body: {r.text}"
    data = r.json()
    assert "url" in data, f"missing 'url' in response: {data}"
    print(f"         Submitted: InitTMA -> SegArrayTMA")
    print(f"         Workflow URL: {data['url']}")


# ---------- service-api ----------

def test_service_api_health():
    r = requests.get(f"{SERVICE_API}/health", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert data.get("status") in ("ok", "healthy"), f"unexpected: {data}"


def test_service_api_openapi():
    r = requests.get(f"{SERVICE_API}/openapi.json", timeout=10)
    assert r.status_code == 200, f"status {r.status_code}"
    data = r.json()
    assert "paths" in data, "missing 'paths' in OpenAPI spec"
    paths = data["paths"]
    assert len(paths) > 0, "no paths in OpenAPI spec"
    print(f"         {len(paths)} endpoints registered:")
    for path in sorted(paths.keys()):
        methods = ", ".join(m.upper() for m in paths[path].keys())
        print(f"           {methods:6s} {path}")


# ---------- main ----------

def main():
    global SIB_MANAGER, EXECUTION_API, SERVICE_API

    parser = argparse.ArgumentParser(description="CincoDeBio endpoint tests")
    parser.add_argument("--sib-manager", default="http://localhost:8081")
    parser.add_argument("--execution-api", default="http://localhost:8082")
    parser.add_argument("--service-api", default="http://localhost:8084")
    args = parser.parse_args()

    SIB_MANAGER = args.sib_manager
    EXECUTION_API = args.execution_api
    SERVICE_API = args.service_api

    print("=== sib-manager tests ===")
    test("health", test_sib_manager_health)
    test("get-sib-map", test_get_sib_map)
    test("get-installed-sibs", test_get_installed_sibs)
    test("get-uninstalled-sibs", test_get_uninstalled_sibs)
    test("get-utd-sib-file", test_get_utd_sib_file)
    test("check-sib-file-hash (invalid)", test_check_sib_file_hash_invalid)
    test("check-sib-files-hashes (empty)", test_check_sib_files_hashes_empty)
    test("sib-manager-state", test_sib_manager_state)

    print("\n=== execution-api tests ===")
    test("health", test_execution_api_health)
    test("get-workflows", test_get_workflows)
    test("submit-workflow", test_submit_workflow)

    print("\n=== service-api tests ===")
    test("health", test_service_api_health)
    test("openapi spec", test_service_api_openapi)

    print(f"\n{'='*40}")
    print(f"Results: {PASSED} passed, {FAILED} failed, {PASSED + FAILED} total")
    sys.exit(1 if FAILED else 0)


if __name__ == "__main__":
    main()
